File: backend/src/OmnesoftChallenge.BLL/Services/ProductService.cs
using ErrorOr;
using OmnesoftChallenge.BLL.Interfaces.Products;
using OmnesoftChallenge.BLL.Mappers;
using OmnesoftChallenge.BLL.Utils;
using OmnesoftChallenge.BLL.Utils.Validators;
using OmnesoftChallenge.BLL.ViewModels.ProductContracts;
using OmnesoftChallenge.DAL.Entities;
using OmnesoftChallenge.DAL.Interfaces.Repository;

namespace OmnesoftChallenge.BLL.Services;

public class ProductService(IProductRepository productRepository) : IProductService
{
    private readonly IProductRepository _productRepository = productRepository;

    public async Task<ErrorOr<ProductListResponse>> GetProducts()
    {
        List<Product> products = await _productRepository.GetAllOrderedById();

        return new ProductListResponse { Products = products.ToProductResponses() };
    }

    public async Task<ErrorOr<ProductResponse>> GetProductById(int id)
    {
        var errors = RequestValidator.ValidateProductId(id);
        if (errors.Count > 0)
        {
            return errors;
        }

        Product? product = await _productRepository.GetById(id);
        if (product is null)
        {
            return ProductNotFound(ErrorCodes.ProductById, id);
        }

        return product.ToProductResponse();
    }

    public async Task<ErrorOr<ProductResponse>> CreateProduct(CreateProductRequest request)
    {
        var errors = RequestValidator.ValidateProduct(request.Name, request.Price, request.Description);
        if (errors.Count > 0)
        {
            return errors;
        }

        var product = request.ToProduct();
        product.InsertionDate = DateTime.UtcNow;
        product.LastUpdateDate = product.InsertionDate;

        product = await _productRepository.Insert(product);

        return product.ToProductResponse();
    }

    public async Task<ErrorOr<ProductResponse>> UpdateProduct(UpdateProductRequest request)
    {
        List<Error> errors =
        [
            .. RequestValidator.ValidateProductId(request.Id),
            .. RequestValidator.ValidateProduct(request.Name, request.Price, request.Description)
        ];
        if (errors.Count > 0)
        {
            return errors;
        }

        Product? product = await _productRepository.GetById(request.Id);
        if (product is null)
        {
            return ProductNotFound(ErrorCodes.UpdateProduct, request.Id);
        }

        request.ApplyTo(product);
        product.LastUpdateDate = DateTime.UtcNow;

        product = await _productRepository.UpdateAsync(product);

        return product.ToProductResponse();
    }

    public async Task<ErrorOr<DeleteProductResponse>> DeleteProduct(int id)
    {
        var errors = RequestValidator.ValidateProductId(id);
        if (errors.Count > 0)
        {
            return errors;
        }

        Product? product = await _productRepository.GetById(id);
        if (product is null)
        {
            return ProductNotFound(ErrorCodes.DeleteProduct, id);
        }

        await _productRepository.Remove(id);

        return new DeleteProductResponse { Id = id, Deleted = true };
    }

    #region Private Helper Methods

    private static Error ProductNotFound(string code, int id)
    {
        return Error.NotFound(
            code: code,
            description: $"Product not found for Id: {id}"
        );
    }

    #endregion
}

File: frontend/src/services/productService.js
import apiClient from './api';
import { getErrorMessage } from '../utils/apiError';

export const getProducts = async () => {
  try {
    const response = await apiClient.get('/products');
    return { success: true, data: response.data.products ?? [] };
  } catch (error) {
    return { success: false, error: getErrorMessage(error, 'Failed to load products.') };
  }
};

export const getProductById = async (id) => {
  try {
    const response = await apiClient.get(`/products/${id}`);
    return { success: true, data: response.data };
  } catch (error) {
    return { success: false, error: getErrorMessage(error, 'Failed to load the product.') };
  }
};

export const createProduct = async (product) => {
  try {
    const response = await apiClient.post('/products', product);
    return { success: true, data: response.data };
  } catch (error) {
    return { success: false, error: getErrorMessage(error, 'Failed to create the product.') };
  }
};

export const updateProduct = async (id, product) => {
  try {
    const response = await apiClient.put(`/products/${id}`, product);
    return { success: true, data: response.data };
  } catch (error) {
    return { success: false, error: getErrorMessage(error, 'Failed to update the product.') };
  }
};

export const deleteProduct = async (id) => {
  try {
    const response = await apiClient.delete(`/products/${id}`);
    return { success: true, data: response.data };
  } catch (error) {
    return { success: false, error: getErrorMessage(error, 'Failed to delete the product.') };
  }
};
